Make Compatible with Windows

IvanAtanasov opened the pull request:
1. BootstrapPython and SetupPython tasks now both execute successfully and perform the same as they do on a linux system.
2. Added RunVenvPython task type. This is designed to make it simple to call python in the virtual environment created with any desired arguments.
3. Added RunVenvPython task type. This allows commands such as conda, pytest etc.. to be run from the virtual environment.

If happy with the approach I can also extend the readme to include these new features.

IvanAtanasov committed:
got bootstrap working on windows

IvanAtanasov committed:
make the conda env creation work on windows

IvanAtanasov committed:
extendable tasks to call executables in venv

IvanAtanasov committed:
refactored and made style checks happy

reverted build.gradle

removed extra blank line

removed comment

IvanAtanasov committed:
included command execution in functional test and added dependecy to all run command tasks

IvanAtanasov committed:
removed blank method

IvanAtanasov committed:
switched type to List

IvanAtanasov committed:
added cmd prefix in setup python

IvanAtanasov committed:
fixed -c to /c

palantirtech commented:
Thanks for your interest in [palantir/gradle-miniconda-plugin](https://github.com/palantir/gradle-miniconda-plugin), @IvanAtanasov! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

mnazbro commented on src/main/java/com/palantir/python/miniconda/tasks/BootstrapPython.java:
```diff
@@ -63,12 +63,18 @@ public void configureAfterEvaluate(final MinicondaExtension miniconda, File cond
         if (os.isWindows()) {
-            executable(condaInstaller);
+            executable("cmd");
+            args("/c");
```
Would it be cleaner to put these all in one args call?

mnazbro commented on src/main/java/com/palantir/python/miniconda/tasks/RunVenvCommand.java:
```diff
@@ -0,0 +1,40 @@
+
+    protected Path getExecutable() {
+        return getMiniconda().getBuildEnvironmentDirectory().toPath()
+                .resolve(getMiniconda().getScriptsRelativeDir() + "/" + executable);
```
This should be possible as two resolves or a resolve to`Paths.get(getMiniconda().getScriptsRelativeDir(), executable)`

mnazbro commented on src/main/java/com/palantir/python/miniconda/tasks/SetupPython.java:
```diff
@@ -61,7 +63,14 @@ public void configureAfterEvaluate(final MinicondaExtension miniconda) {
         getOutputs().dir(miniconda.getBuildEnvironmentDirectory());
-        executable(miniconda.getBootstrapDirectory().toPath().resolve("bin/conda"));
+        Path condaPath = miniconda.getBootstrapDirectory().toPath()
+                .resolve(miniconda.getScriptsRelativeDir() + "/conda");
```
Same comment as before, probably worth using only paths rather than using slashes.

mnazbro reviewed (commented):
Thanks for your pull request @IvanAtanasov 

I have some minor comments. I'll do another review after I do a quick smoke test on a Windows computer today.

IvanAtanasov committed:
Use paths rather than strings representing paths. Also allow env variables to be set when running commands

IvanAtanasov commented:
Thanks for taking a look @mnazbro. Hopefully this commit has addressed your comments. It also adds the ability to define environment variables when running conda/python commands.

Did you ever try it on windows? Did it work for you?

mnazbro commented:
I tested this out on Windows and was having some Path issues, so I'll do a bit more investigation into whether I was using it wrong.

IvanAtanasov committed:
resolve merge conflicts

IvanAtanasov commented:
@mnazbro I've resolved the merge conflicts introduced by recent merges. Could you have another look please? I think there are going to be extra changes required in order to make the conda build functionality work on windows however.
